kubeflow/trainer/rhai/instrumentation/traininghub_progression.py

```
@@ -0,0 +1,683 @@
+            LoRA writes a single metrics file (no per-rank wildcard).
+            Each line is a JSON object with step, epoch, loss, learning_rate.
+            """
+            metrics_file = f"{ckpt_output_dir}/{metrics_file_rank0}"
+
+            try:
+                if not os.path.exists(metrics_file):
+                    return {}
+
+                # Read last line of metrics using tail
+                try:
+                    result = subprocess.run(
```

Ruff (0.16.0)

[error] `subprocess` call: check for execution of untrusted input

(S603)